### ISSUE TITLE
[i18n] Some language aren't support by i18n

### ISSUE THREAD

**Issue by JulienHe** [VirtusLab-Open-Source/strapi-plugin-navigation#335]
**TLDR**: Can't update the navigation title in Japanese. [Video](https://share.cleanshot.com/QW9W85Zm).

## What

While working on a project for a client, we need to have Strapi handle the Japanese language.
The update isn't happening when you click on Save.

The payload was: `q: %E6%88%A6%E7%95%A5%E3%83%AD%E3%83%BC%E3%83%89%E3%83%9E%E3%83%83%E3%83%97`. 
I did try with different languages like Korean and had the same problem.

One interesting thing to notice is that if you add a Romanian letter in the title, the update is happening. Beginning, end or in the title itself.
<img width="841" alt="CleanShot 2023-05-14 at 11 41 00@2x" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/718868/72048f33-e819-4a4c-b979-549ee9eba80e">

<img width="810" alt="CleanShot 2023-05-14 at 11 42 23@2x" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/718868/2a8eb8e5-29ad-44c8-86fd-3fcebabce448">

<img width="811" alt="CleanShot 2023-05-14 at 11 44 12@2x" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/718868/479a04aa-9b85-42bb-893a-28ce3df9a6d8">

### Comparative table
I run some tests to see which language had issues (as far as possible)
Native = only the character from that specific language
Mix = With one or more Romanian letters (A-Z)

| Language  |  Native | Mix  |
|---|---|---|
|  Japanese |  🚫 | ✅  |
| Korean  |  🚫 | ✅  |
| Arabic  |  ✅ | ✅  |
| Amharic  |  🚫 | ✅  |
| Armenian  |  ✅ | ✅  |
| Assamese  |  🚫 | ✅  |
| Bengali  |  🚫 | ✅  |
| Bhojpuri  |  🚫 | ✅  |
| Chinese (Simplified)  |  🚫 | ✅  |
| Chinese (Traditional)  |  🚫 | ✅  |
| Dhiveri  |  ✅ | ✅  |
| Dogri |  🚫 | ✅  |
| Georgian |  ✅ | ✅  |
| Greek |  ✅ | ✅  |
| Gujarati |  🚫 | ✅  |
| Hebrew |  🚫 | ✅  |
| Hindi |  🚫 | ✅  |
| Japanese |  🚫 | ✅  |
| Kannada |  🚫 | ✅  |
| Khmer |  🚫 | ✅  |
| Konkani |  🚫 | ✅  |
| Korean |  🚫 | ✅  |
| Kurdish |  ✅ | ✅  |
| Lao |  🚫 | ✅  |
| Maithili |  🚫 | ✅  |
| Malayalam |  🚫 | ✅  |
| Marathi |  🚫 | ✅  |
| Manipuri |  🚫 | ✅  |
| Myanmar |  🚫 | ✅  |
| Nepali |  🚫 | ✅  |
| Odia |  🚫 | ✅  |
| Pashto |  ✅ | ✅  |
| Persian |  ✅ | ✅  |
| Punjabi |  🚫 | ✅  |
| Sanskrit |  🚫 | ✅  |
| Sindhi |  ✅ | ✅  |
| Sinhala |  🚫 | ✅  |
| Tamil |  🚫 | ✅  |
| Telugu |  🚫 | ✅  |
| Thai |  🚫 | ✅  |
| Tigrinya |  🚫 | ✅  |
| Urdu |  ✅| ✅  |
| Uyghur |  ✅| ✅  |
| Yiddish |  🚫 | ✅  |

**Comment by ZacharyChim** [comment]
I have the same issue. If you have only Chinese (Traditional) characters, the Save button won't work. Any idea?

[example](https://d38xoolefqxt5u.cloudfront.net/original/2X/2/248b3df74fa0cb7195e8150983105db758beb799.gif)

**Comment by JulienHe** [comment]
@ZacharyChim Yeah, I listed Chinese (traditional and Simplified) in the missing table too.

**Comment by cyp3rius** [comment]
Thanks for such detailed check. I'm moving this issue to investigation now.

**Comment by cyp3rius** [comment]
So in reality... it's not an issue :) It's some kind of consequence of the approach we took in a past with opening and simplifying the `getSlug` service method.

If you follow this [chapter od documentation](https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation#slug-generation) everything is going to work. Edit the `preprocess` by the way which is going to allow you to slugify non AZ characters. Example which worked for me (is not perfect as it encoding characters in slug) is:

```ts
const preprocess = (q) => encodeURIComponent(q);
```

**Comment by ZacharyChim** [comment]
Thank you for your response, @cyp3rius 

I'm rather new to Strapi and I followed the link you provided, but can't find `/admin/src/index.js`, but only `node_modules/strapi-plugin-navigation/admin/src/index.js`. In that file, I did find the `bootstrap` section, but after adding the code:

```
bootstrap({ strapi }) {
        const navigationCommonService = strapi.plugin("navigation").service("common");
        const originalGetSlug = navigationCommonService.getSlug;
        const preprocess = (q) => {
          return encodeURIComponent(q);
        };
    
        navigationCommonService.getSlug = (query) => {
          return originalGetSlug(preprocess(query));
        };
      },
```

And re-build Strapi, run yarn dev, Strapi turns blank, with the error:

```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'plugin')
    at bootstrap (main.10fb271a.js:908:5663)
    at admin-app.083a7a37.chunk.js:66:7279
    at Array.forEach (<anonymous>)
    at Zs.bootstrap (admin-app.083a7a37.chunk.js:66:7227)
    at main.10fb271a.js:417:41287
```

Am I doing wrong? Thanks

**Comment by cyp3rius** [comment]
@ZacharyChim you should modify that in your Strapi project, not the plugin files. The bootstrap method is provided as empty by Strapi in default configuration.

**Comment by cyp3rius** [comment]
That's the correct file.

Should looks like this one:

```ts
'use strict';

module.exports = {
  /**
   * An asynchronous register function that runs before
   * your application is initialized.
   *
   * This gives you an opportunity to extend code.
   */
  register(/*{ strapi }*/) {},

  /**
   * An asynchronous bootstrap function that runs before
   * your application gets started.
   *
   * This gives you an opportunity to set up your data model,
   * run jobs, or perform some special logic.
   */
  bootstrap({ strapi }) {
    const navigationCommonService = strapi.plugin("navigation").service("common");
    const originalGetSlug = navigationCommonService.getSlug;
    const preprocess = (q) => encodeURIComponent(q);
  
    navigationCommonService.getSlug = (query) => {
      return originalGetSlug(preprocess(query));
    };
  },
};
```

**Comment by ZacharyChim** [comment]
Hi @cyp3rius 

I updated the file, re-build, although the build is successful, there's an error:

```
Error: write EPIPE
    at ChildProcess.target._send (node:internal/child_process:867:20)
    at ChildProcess.target.send (node:internal/child_process:740:19)
    at Worker.send (node:internal/cluster/worker:48:10)
    at EventEmitter.<anonymous> (.../backend/node_modules/@strapi/strapi/lib/commands/develop.js:92:16)
```

When I run the build command again, the error disappeared, and then I run `run dev`, it's the same, I can't save the Chinese characters. Any idea?

[screenshot](https://imgur.com/a/o1VTU9r)

**Comment by ZacharyChim** [comment]
Anyone got any idea? Thanks

**Comment by cyp3rius** [comment]
@ZacharyChim I've tried to reproduce your error without a luck. IF you remove the body of `boostrap` you can run project without any issues?

**Comment by JulienHe** [comment]
Did some updates and the example you provided works for me, thanks! 
As you'll probably get more and more people using the plugin, what do you think of adding some instructions in the documentation to handle i18n better? Might be helpful.

**Comment by eigengrau-ch** [comment]
Hi @CodeVoyager

Sorry for the late reply, i didn't saw that you posted here. Thank you for handling my request, i will try to provide you with all the information necessary.

- Not sure what you exactly mean by generating `slug` from `Title`, but we don't do that.
- This is only half true for me. Some non-latin characters like krylic or arabic characters do work, but chinese don't. Check the below video to see how it behaves on my end (v2.2.12)
- Yes, but this seems to be only true for the `Url` field but not for the `Title` field inside the nav item popup form
- Thats true. It seems some non-latin characters fail in saving without throwing a error nor a error message on the field ([see video](https://www.loom.com/share/6ffbef4db7dc4699ad9194b5c7bc15c0))

Based on my testings, it seems that not all non-latin characters work. Watching the language table on the issue description here. Obviously I have not tested all of those not working languages, but I can confirm that chinese characters don't work.

Yes. My question was, if there is a solution to apply the same kind of solution provided by [cyp3rius post](https://github.com/cyp3rius) but for `Title` instead of `Url`.

**Comment by CodeVoyager** [comment]
Hi @eigengrau-ch 

No worries, let's got through all points.

- By `slug` I meant `uiRouterKey` field in a menu response. See image below

<img width="1049" alt="image" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/13495487/17b86b76-ee54-49b1-aaac-5de34fb395a3">

- In order to save an item a call to server is made to transform `title` to `uiRouterKey`. Built-in solution works for latin characters and some non-latin characters. Any character that is unrecognisable is ignored. If all of title is unreasonable empty string is a result. Empty `uiRouterKey` is invalid so item cannot be saved. Action point for us is to present this validation message better. In screenshot below you can see result of said `uiRouterKey` generation on my instance. It's result for string "晨鳥". Note that my instance uses custom slug generation. The same one that @cyp3rius provided.

<img width="933" alt="image" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/13495487/dcaf613e-b971-4478-969b-e979e0867812">

- Can you confirm that in file `src/index.js` in method `bootstrap()` `getSlug()` from navigation service is overridden with a custom solution? For example:

<img width="508" alt="image" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/13495487/ec52c5ff-8b97-436e-bf6a-14841f3da299">

**Comment by eigengrau-ch** [comment]
Hi @CodeVoyager, thank you very much for your reply. I am very sorry, you were right. Not knowing, i should have used the script inside the `bootstrap` of `src/admin/app.js` instead of `src/index.js`. Somehow this worked for the `Url` but not for `Title` field, leading me to think it worked partially. Also the previous posts suggested to add it inside of `src/admin/app.js`.

Thanks for supporting and investigating into my issue :)

**Comment by cyp3rius** [comment]
Implemented in `v2.2.13`

**Comment by suyumin82** [comment]
Hi @cyp3rius and @CodeVoyager i am new to strapi and we are building chinese simplify CMS leverage on strapi with i18n. Without changing much of the code, i am getting internal error when adding chinese characters. Based on the comment above, does it meant i need to implement slug? does it not work with out of the box strapi?
FYI i am on strapi v4.15.5
node: v18.19.0
<img width="760" alt="image" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/11375598/21f86fcf-73cf-4e0f-9e36-f2d829afd4f5">

<img width="402" alt="image" src="https://github.com/VirtusLab-Open-Source/strapi-plugin-navigation/assets/11375598/7f97e198-5333-4fff-b4d8-e7321f5c7135">

appreciate your help on this.
